Build failure: git-branchless (Rust)

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

build `git-branchless`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/305753849

### Relevant log output

Can’t find how to disable ANSI escapes for this in 2 minutes, so pasting with

```console
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/5mnrxasmz579iigavsz854xgj4myslw3-source
source root is source
Executing cargoSetupPostUnpackHook
Finished cargoSetupPostUnpackHook
unpackPhase completed in 1 minutes 19 seconds
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Executing cargoSetupPostPatchHook
Validating consistency between /build/source/Cargo.lock and /build/git-branchless-0.10.0-vendor/Cargo.lock
Finished cargoSetupPostPatchHook
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing cargoBuildHook
cargoBuildHook flags: -j 16 --target x86_64-unknown-linux-gnu --offline --profile release
[1m[32m   Compiling[0m libc v0.2.153
[1m[32m   Compiling[0m proc-macro2 v1.0.79
[1m[32m   Compiling[0m unicode-ident v1.0.12
[1m[32m   Compiling[0m cfg-if v1.0.0
[1m[32m   Compiling[0m autocfg v1.2.0
[1m[32m   Compiling[0m once_cell v1.19.0
[1m[32m   Compiling[0m log v0.4.21
[1m[32m   Compiling[0m version_check v0.9.4
[1m[32m   Compiling[0m shlex v1.3.0
[1m[32m   Compiling[0m memchr v2.7.2
[1m[32m   Compiling[0m parking_lot_core v0.9.9
[1m[32m   Compiling[0m smallvec v1.13.2
[1m[32m   Compiling[0m ident_case v1.0.1
[1m[32m   Compiling[0m crossbeam-utils v0.8.19
[1m[32m   Compiling[0m pin-project-lite v0.2.14
[1m[32m   Compiling[0m zerocopy v0.7.32
[1m[32m   Compiling[0m fnv v1.0.7
[1m[32m   Compiling[0m lazy_static v1.5.0
[1m[32m   Compiling[0m unicode-width v0.1.14
[1m[32m   Compiling[0m regex-syntax v0.8.3
[1m[32m   Compiling[0m either v1.11.0
[1m[32m   Compiling[0m pkg-config v0.3.30
[1m[32m   Compiling[0m ahash v0.8.11
[1m[32m   Compiling[0m bitflags v1.3.2
[1m[32m   Compiling[0m allocator-api2 v0.2.18
[1m[32m   Compiling[0m itoa v1.0.11
[1m[32m   Compiling[0m lock_api v0.4.11
[1m[32m   Compiling[0m num-traits v0.2.18
[1m[32m   Compiling[0m tracing-core v0.1.32
[1m[32m   Compiling[0m bitflags v2.5.0
[1m[32m   Compiling[0m serde v1.0.204
[1m[32m   Compiling[0m scopeguard v1.2.0
[1m[32m   Compiling[0m vcpkg v0.2.15
[1m[32m   Compiling[0m signal-hook v0.3.17
[1m[32m   Compiling[0m quote v1.0.36
[1m[32m   Compiling[0m rustversion v1.0.15
[1m[32m   Compiling[0m utf8parse v0.2.1
[1m[32m   Compiling[0m ppv-lite86 v0.2.17
[1m[32m   Compiling[0m regex-syntax v0.6.29
[1m[32m   Compiling[0m syn v2.0.58
[1m[32m   Compiling[0m thiserror v1.0.63
[1m[32m   Compiling[0m heck v0.5.0
[1m[32m   Compiling[0m aho-corasick v1.1.3
[1m[32m   Compiling[0m crossbeam-channel v0.5.12
[1m[32m   Compiling[0m thread_local v1.1.8
[1m[32m   Compiling[0m static_assertions v1.1.0
[1m[32m   Compiling[0m getrandom v0.1.16
[1m[32m   Compiling[0m stable_deref_trait v1.2.0
[1m[32m   Compiling[0m pin-utils v0.1.0
[1m[32m   Compiling[0m anstyle v1.0.6
[1m[32m   Compiling[0m unicode-segmentation v1.11.0
[1m[32m   Compiling[0m typenum v1.17.0
[1m[32m   Compiling[0m crossbeam-epoch v0.9.18
[1m[32m   Compiling[0m num-rational v0.4.1
[1m[32m   Compiling[0m slab v0.4.9
[1m[32m   Compiling[0m generic-array v0.14.7
[1m[32m   Compiling[0m jobserver v0.1.30
[1m[32m   Compiling[0m powerfmt v0.2.0
[1m[32m   Compiling[0m maybe-uninit v2.0.0
[1m[32m   Compiling[0m rustix v0.38.32
[1m[32m   Compiling[0m crossbeam-deque v0.8.5
[1m[32m   Compiling[0m deranged v0.3.11
[1m[32m   Compiling[0m tracing-log v0.1.4
[1m[32m   Compiling[0m indexmap v1.9.3
[1m[32m   Compiling[0m memoffset v0.6.5
[1m[32m   Compiling[0m cc v1.1.28
[1m[32m   Compiling[0m getrandom v0.2.14
[1m[32m   Compiling[0m signal-hook-registry v1.4.1
[1m[32m   Compiling[0m mio v0.8.11
[1m[32m   Compiling[0m num-integer v0.1.46
[1m[32m   Compiling[0m parking_lot v0.12.1
[1m[32m   Compiling[0m rand_core v0.6.4
[1m[32m   Compiling[0m sharded-slab v0.1.7
[1m[32m   Compiling[0m hashbrown v0.14.3
[1m[32m   Compiling[0m time-core v0.1.2
[1m[32m   Compiling[0m crunchy v0.2.2
[1m[32m   Compiling[0m num_threads v0.1.7
[1m[32m   Compiling[0m ansi_term v0.12.1
[1m[32m   Compiling[0m futures-sink v0.3.31
[1m[32m   Compiling[0m futures-core v0.3.31
[1m[32m   Compiling[0m tinyvec_macros v0.1.1
[1m[32m   Compiling[0m linux-raw-sys v0.4.13
[1m[32m   Compiling[0m iana-time-zone v0.1.60
[1m[32m   Compiling[0m paste v1.0.14
[1m[32m   Compiling[0m signal-hook-mio v0.2.3
[1m[32m   Compiling[0m num-conv v0.1.0
[1m[32m   Compiling[0m ryu v1.0.17
[1m[32m   Compiling[0m chrono v0.4.34
[1m[32m   Compiling[0m time v0.3.36
[1m[32m   Compiling[0m futures-channel v0.3.31
[1m[32m   Compiling[0m tinyvec v1.6.0
[1m[32m   Compiling[0m regex-automata v0.4.6
[1m[32m   Compiling[0m regex-automata v0.1.10
[1m[32m   Compiling[0m num-iter v0.1.44
[1m[32m   Compiling[0m rand_chacha v0.3.1
[1m[32m   Compiling[0m rand_core v0.5.1
[1m[32m   Compiling[0m num-complex v0.4.5
[1m[32m   Compiling[0m memmap v0.7.0
[1m[32m   Compiling[0m libz-sys v1.1.16
[1m[32m   Compiling[0m castaway v0.2.2
[1m[32m   Compiling[0m anstyle-parse v0.2.3
[1m[32m   Compiling[0m itertools v0.13.0
[1m[32m   Compiling[0m itertools v0.12.1
[1m[32m   Compiling[0m fastrand v2.0.2
[1m[32m   Compiling[0m colorchoice v1.0.0
[1m[32m   Compiling[0m futures-task v0.3.31
[1m[32m   Compiling[0m matchers v0.1.0
[1m[32m   Compiling[0m futures-io v0.3.31
[1m[32m   Compiling[0m hashbrown v0.12.3
[1m[32m   Compiling[0m anyhow v1.0.86
[1m[32m   Compiling[0m rayon-core v1.12.1
[1m[32m   Compiling[0m anstyle-query v1.0.2
[1m[32m   Compiling[0m serde_json v1.0.120
[1m[32m   Compiling[0m anstream v0.6.13
[1m[32m   Compiling[0m darling_core v0.20.8
[1m[32m   Compiling[0m unicode-normalization v0.1.23
[1m[32m   Compiling[0m num v0.4.2
[1m[32m   Compiling[0m compact_str v0.7.1
[1m[32m   Compiling[0m smallvec v0.6.14
[1m[32m   Compiling[0m backtrace v0.3.71
[1m[32m   Compiling[0m libgit2-sys v0.17.0+1.8.1
[1m[32m   Compiling[0m libsqlite3-sys v0.26.0
[1m[32m   Compiling[0m rand v0.8.5
[1m[32m   Compiling[0m unicode-truncate v1.0.0
[1m[32m   Compiling[0m tempfile v3.10.1
[1m[32m   Compiling[0m rand_chacha v0.2.2
[1m[32m   Compiling[0m lru v0.12.3
[1m[32m   Compiling[0m crossterm v0.27.0
[1m[32m   Compiling[0m serial-core v0.4.0
[1m[32m   Compiling[0m tracing-attributes v0.1.27
[1m[32m   Compiling[0m serde_derive v1.0.204
[1m[32m   Compiling[0m thiserror-impl v1.0.63
[1m[32m   Compiling[0m strum_macros v0.26.4
[1m[32m   Compiling[0m futures-macro v0.3.31
[1m[32m   Compiling[0m enum-map-derive v0.17.0
[1m[32m   Compiling[0m regex v1.10.5
[1m[32m   Compiling[0m stability v0.2.0
[1m[32m   Compiling[0m rand v0.4.6
[1m[32m   Compiling[0m enum-map v2.7.3
[1m[32m   Compiling[0m termios v0.2.2
[1m[32m   Compiling[0m ioctl-rs v0.1.6
[1m[32m   Compiling[0m owning_ref v0.4.1
[1m[32m   Compiling[0m crossbeam-queue v0.3.11
[1m[32m   Compiling[0m vte_generate_state_changes v0.1.1
[1m[32m   Compiling[0m itertools v0.10.5
[1m[32m   Compiling[0m clap_lex v0.7.0
[1m[32m   Compiling[0m futures-util v0.3.31
[1m[32m   Compiling[0m arrayvec v0.7.4
[1m[32m   Compiling[0m eyre v0.6.12
[1m[32m   Compiling[0m darling_macro v0.20.8
[1m[32m   Compiling[0m adler v1.0.2
[1m[32m   Compiling[0m unicode-linebreak v0.1.5
[1m[32m   Compiling[0m doc-comment v0.3.3
[1m[32m   Compiling[0m xi-unicode v0.3.0
[1m[32m   Compiling[0m cassowary v0.3.0
[1m[32m   Compiling[0m byteorder v1.5.0
[1m[32m   Compiling[0m portable-atomic v1.6.0
[1m[32m   Compiling[0m tiny-keccak v2.0.2
[1m[32m   Compiling[0m unicode-bidi v0.3.15
[1m[32m   Compiling[0m smawk v0.3.2
[1m[32m   Compiling[0m gimli v0.28.1
[1m[32m   Compiling[0m predicates-core v1.0.6
[1m[32m   Compiling[0m esl01-vlqencoding v0.3.0
[1m[32m   Compiling[0m strsim v0.11.1
[1m[32m   Compiling[0m percent-encoding v2.3.1
[1m[32m   Compiling[0m strum v0.26.2
[1m[32m   Compiling[0m tracing v0.1.40
[1m[32m   Compiling[0m ratatui v0.27.0
[1m[32m   Compiling[0m darling v0.20.8
[1m[32m   Compiling[0m clap_builder v4.5.9
[1m[32m   Compiling[0m enumset_derive v0.8.1
[1m[32m   Compiling[0m form_urlencoded v1.2.1
[1m[32m   Compiling[0m tracing-subscriber v0.3.11
[1m[32m   Compiling[0m esl01-atomicfile v0.3.0
[1m[32m   Compiling[0m idna v0.5.0
[1m[32m   Compiling[0m textwrap v0.16.1
[1m[32m   Compiling[0m parking_lot_core v0.2.14
[1m[32m   Compiling[0m miniz_oxide v0.7.2
[1m[32m   Compiling[0m vte v0.11.1
[1m[32m   Compiling[0m crossbeam v0.8.4
[1m[32m   Compiling[0m serial-unix v0.4.0
[1m[32m   Compiling[0m clap_derive v4.5.8
[1m[32m   Compiling[0m twox-hash v1.6.3
[1m[32m   Compiling[0m enumset v1.1.3
[1m[32m   Compiling[0m rand v0.7.3
[1m[32m   Compiling[0m cursive_core v0.3.7
[1m[32m   Compiling[0m nix v0.25.1
[1m[32m   Compiling[0m tracing-error v0.2.0
[1m[32m   Compiling[0m addr2line v0.21.0
[1m[32m   Compiling[0m futures-executor v0.3.30
[1m[32m   Compiling[0m block-buffer v0.10.4
[1m[32m   Compiling[0m crypto-common v0.1.6
[1m[32m   Compiling[0m is-terminal v0.4.12
[1m[32m   Compiling[0m fs2 v0.4.3
[1m[32m   Compiling[0m owning_ref v0.3.3
[1m[32m   Compiling[0m object v0.32.2
[1m[32m   Compiling[0m rustc-demangle v0.1.23
[1m[32m   Compiling[0m owo-colors v3.5.0
[1m[32m   Compiling[0m indenter v0.3.3
[1m[32m   Compiling[0m difflib v0.4.0
[1m[32m   Compiling[0m hex v0.4.3
[1m[32m   Compiling[0m termtree v0.4.1
[1m[32m   Compiling[0m assert_cmd v2.0.14
[1m[32m   Compiling[0m predicates-tree v1.0.9
[1m[32m   Compiling[0m color-spantrace v0.2.1
[1m[32m   Compiling[0m predicates v3.1.0
[1m[32m   Compiling[0m bytes v1.7.2
[1m[32m   Compiling[0m esl01-mincode v0.3.0
[1m[32m   Compiling[0m esl01-minibytes v0.3.0
[1m[32m   Compiling[0m esl01-dag-types v0.3.0
[1m[32m   Compiling[0m scm-record v0.4.0
[1m[32m   Compiling[0m esl01-indexedlog v0.3.0
[0m[1m[38;5;9merror[E0308][0m[0m[1m: `match` arms have incompatible types[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0m/build/git-branchless-0.10.0-vendor/esl01-indexedlog-0.3.0/src/lock.rs:138:30[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m134[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m/[0m[0m [0m[0m        match (opts.exclusive, opts.non_blocking) {[0m
[0m[1m[38;5;12m135[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m            (true, false) => file.lock_exclusive(),[0m
[0m    [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m                              [0m[0m[1m[38;5;12m---------------------[0m[0m [0m[0m[1m[38;5;12mthis is found to be of type `std::result::Result<(), std::io::Error>`[0m
[0m[1m[38;5;12m136[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m            (true, true) => file.try_lock_exclusive(),[0m
[0m    [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m                             [0m[0m[1m[38;5;12m-------------------------[0m[0m [0m[0m[1m[38;5;12mthis is found to be of type `std::result::Result<(), std::io::Error>`[0m
[0m[1m[38;5;12m137[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m            (false, false) => file.lock_shared(),[0m
[0m    [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m                               [0m[0m[1m[38;5;12m------------------[0m[0m [0m[0m[1m[38;5;12mthis is found to be of type `std::result::Result<(), std::io::Error>`[0m
[0m[1m[38;5;12m138[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m            (false, true) => file.try_lock_shared(),[0m
[0m    [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m                              [0m[0m[1m[38;5;9m^^^^^^^^^^^^^^^^^^^^^^[0m[0m [0m[0m[1m[38;5;9mexpected `Result<(), Error>`, found `Result<(), TryLockError>`[0m
[0m[1m[38;5;12m139[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m        }[0m
[0m    [0m[0m[1m[38;5;12m|[0m[0m [0m[0m[1m[38;5;12m|_________-[0m[0m [0m[0m[1m[38;5;12m`match` arms have incompatible types[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m    [0m[0m[1m[38;5;12m= [0m[0m[1mnote[0m[0m: expected enum `std::result::Result<_, std::[0m[0m[1m[35mio::Error[0m[0m>`[0m
[0m               found enum `std::result::Result<_, std::[0m[0m[1m[35mfs::TryLockError[0m[0m>`[0m

[1m[32m   Compiling[0m clap v4.5.9
[1m[32m   Compiling[0m rayon v1.10.0
[0m[1mFor more information about this error, try `rustc --explain E0308`.[0m
[1m[31merror[0m[1m:[0m could not compile `esl01-indexedlog` (lib) due to 1 previous error
[1m[33mwarning[0m[1m:[0m build failed, waiting for other jobs to finish...
```

### Additional context

_No response_

### System metadata

```sh-session
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.42, NixOS, 25.11 (Xantusia), 25.11.20250819.2007595`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - nixpkgs: `/nix/store/pxnx7dbr0pz3qpgw2r315wdbb00afdw7-source`
```

### Notify maintainers

@nh2 @hmenke @bryango 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

[^1]

[^1]: Please consider [giving up MS GitHub](https://sfconservancy.org/GiveUpGitHub/) or offering a non-proprietary, non-US-corporate-controlled mirror for this free software project. I wish to delete this Microsoft account in the future, but I need more projects like this to support alternative methods to send patches & contribute.

--- a/pkgs/by-name/gi/git-branchless/fix-esl01-indexedlog-for-rust-1_89.patch
+++ b/pkgs/by-name/gi/git-branchless/fix-esl01-indexedlog-for-rust-1_89.patch
@@ -1,0 +1,18 @@
+diff --git a/src/lock.rs b/src/lock.rs
+--- a/src/lock.rs
++++ b/src/lock.rs
+@@ -132,10 +132,10 @@ impl ScopedDirLock {
+ 
+         // Lock
+         match (opts.exclusive, opts.non_blocking) {
+-            (true, false) => file.lock_exclusive(),
+-            (true, true) => file.try_lock_exclusive(),
+-            (false, false) => file.lock_shared(),
+-            (false, true) => file.try_lock_shared(),
++            (true, false) => fs2::FileExt::lock_exclusive(&file),
++            (true, true) => fs2::FileExt::try_lock_exclusive(&file),
++            (false, false) => fs2::FileExt::lock_shared(&file),
++            (false, true) => fs2::FileExt::try_lock_shared(&file),
+         }
+         .context(&path, || {
+             format!(

--- a/pkgs/by-name/gi/git-branchless/package.nix
+++ b/pkgs/by-name/gi/git-branchless/package.nix
@@ -21,6 +21,20 @@ rustPlatform.buildRustPackage rec {
     hash = "sha256-8uv+sZRr06K42hmxgjrKk6FDEngUhN/9epixRYKwE3U=";
   };
 
+  # Patch the vendored esl01-indexedlog crate.
+  # This is necessary to fix the build for rust 1.89. See:
+  # - https://github.com/NixOS/nixpkgs/issues/437051
+  # - https://github.com/arxanas/git-branchless/issues/1585
+  # - https://github.com/facebook/sapling/issues/1119
+  # The patch is derived from:
+  # - https://github.com/facebook/sapling/commit/9e27acb84605079bf4e305afb637a4d6801831ac
+  postPatch = ''
+    (
+      cd ../git-branchless-*-vendor/esl01-indexedlog-*/
+      patch -p1 < ${./fix-esl01-indexedlog-for-rust-1_89.patch}
+    )
+  '';
+
   cargoHash = "sha256-i7KpTd4fX3PrhDjj3R9u98rdI0uHkpQCxSmEF+Gu7yk=";
 
   nativeBuildInputs = [ pkg-config ];

